src/Common/FileCache.h

```diff
@@ -0,0 +1,182 @@
+{
+
+/**
+ * Local cache for remote filesystem files, represented as a set of non-overlapping non-empty file segments.
+ */
+class FileCache : boost::noncopyable
+{
+friend class FileSegment;
+
+public:
+    using Key = UInt128;
+    using Downloader = std::unique_ptr<SeekableReadBuffer>;
+
+    FileCache(
+        const String & cache_base_path_,
+        size_t max_size_, size_t max_element_size_);
+
+    virtual ~FileCache() = default;
+
+    static bool shouldBypassCache();
+
+    size_t capacity() const { return max_size; }
+
+    static Key hash(const String & path);
+
+    String path(const Key & key, size_t offset);
+
+    String path(const Key & key);
+
+    /**
+     * Given an `offset` and `size` representing [offset, offset + size) bytes interval,
+     * return list of cached non-overlapping non-empty
+     * file segments `[segment1, ..., segmentN]` which intersect with given interval.
+     *
+     * Segments in returned list are ordered in ascending order and represent a full contiguous
+     * interval (no holes). Each segment in returned list has state: DOWNLOADED, DOWNLOADING or EMPTY.
+     *
+     * As long as pointers to returned file segments are hold
+     * it is guaranteed that these file segments are not removed from cache.
+     */
+    virtual FileSegmentsHolder getOrSet(const Key & key, size_t offset, size_t size) = 0;
+
+    virtual void remove(const Key & key) = 0;
+
+    /// For debug.
+    virtual String dumpStructure(const Key & key) = 0;
+
+protected:
+    String cache_base_path;
+    size_t max_size = 0;
+    size_t max_element_size = 0;
+
+    mutable std::mutex mutex;
+
+    virtual bool tryReserve(
+        const Key & key, size_t offset, size_t size,
+        std::lock_guard<std::mutex> & cache_lock) = 0;
+
+    virtual void remove(
+        Key key, size_t offset,
+        std::lock_guard<std::mutex> & cache_lock) = 0;
+
+    virtual bool isLastFileSegmentHolder(
+        const Key & key, size_t offset,
+        std::lock_guard<std::mutex> & cache_lock) = 0;
+
+    virtual void reduceSizeToDownloaded(
+        const Key & key, size_t offset, std::lock_guard<std::mutex> & cache_lock) = 0;
+};
+
+using FileCachePtr = std::shared_ptr<FileCache>;
+
+class LRUFileCache final : public FileCache
+{
+public:
+    LRUFileCache(const String & cache_base_path_, size_t max_size_, size_t max_element_size_ = 0);
+
+    FileSegmentsHolder getOrSet(const Key & key, size_t offset, size_t size) override;
+
+    void remove(const Key & key) override;
+
+private:
+    using FileKeyAndOffset = std::pair<Key, size_t>;
+    using LRUQueue = std::list<FileKeyAndOffset>;
+    using LRUQueueIterator = typename LRUQueue::iterator;
+
+    struct FileSegmentCell : boost::noncopyable
+    {
+        FileSegmentPtr file_segment;
+
+        /// Iterator is put here on first reservation attempt, if successful.
+        std::optional<LRUQueueIterator> queue_iterator;
+
+        bool releasable() const { return file_segment.unique(); }
```

Comment:
Quite dangerous to use shared_ptr internals for tracking something... Need to a comment who can hold the pointer and when it's possible to release it.